llama_index/indices/query/embedding_utils.py

```diff
@@ -92,3 +92,62 @@ def get_top_k_embeddings_learner(
     result_ids = [embedding_ids[ix] for ix in top_sorted_ix]
 
     return result_similarities, result_ids
+
+
+def get_top_k_mmr_embeddings(
+    query_embedding: List[float],
+    embeddings: List[List[float]],
+    similarity_fn: Optional[Callable[..., float]] = None,
+    similarity_top_k: Optional[int] = None,
+    embedding_ids: Optional[List] = None,
+    similarity_cutoff: Optional[float] = None,
+    mmr_threshold: Optional[float] = None,
+) -> Tuple[List[float], List]:
+    """Get top nodes by similarity to the query, 
+    discount by their similarity to previous results.
+
+    A mmr_threshold of 0 will strongly avoid similarity to previous results.
+    A mmr_threshold of 1 will compare similarity the query and ignore previous results.
+
+    """
+
+    threshold = mmr_threshold or 0.5
+    if embedding_ids is None:
+        embedding_ids = [i for i in range(len(embeddings))]
+    similarity_fn = similarity_fn or default_similarity_fn
+    embed_map = dict(zip(embedding_ids, range(len(embedding_ids))))
+    similarities: List[List[Any]] = (
+        []
+    )  
+    # similarities is omposed of lists of similarity score against query, id, 
+    # and similarity score against other results
+
+    for i, emb in enumerate(embeddings):
+        similarity = similarity_fn(query_embedding, emb)
+        similarities.append([similarity, embedding_ids[i], 0])
```

Comment:
nit: can you use a tuple here instead? `(similarity, embedding_ids[i], 0)`

Reply:
The reason I'm using a tuple is that the third entry, the zero, is the diversity score and is updated. Tuples don't support item assignment.

Reply:
Right!